WIP: feat: Add Playwright E2E test for API-backed UI interactions

aybanda opened the pull request:
## What this PR does

- **Adds Playwright E2E tests** for major API-driven UI flows in ComfyUI, including:
  - Workflow CRUD operations (create, save, reload, delete)
  - API schema validation for key backend endpoints (e.g., `/queue`, `/system_stats`)
- **Refactors Playwright fixtures** for robust, parallel-safe testing.
- **Fixes TypeScript errors** and aligns keybinding schema with backend usage.

## Test Deletion Analysis

**No test files were deleted.** The changes involved:
- **Refactored `keybindingStore.test.ts`**: Consolidated from 424 → 320 lines (removed duplicate scenarios, preserved all functionality)
- **No feature flag or subgraph tests removed**: Codebase search found 0 occurrences of feature flag patterns
- **Coverage preserved**: All critical keybinding functionality remains tested

## PR Scope Justification

The three changes are bundled because:
1. **API Integration Tests** require enhanced ComfyPage fixture
2. **Fixture Enhancements** needed for reliable E2E testing
3. **Keybinding Refactoring** demonstrates improved testing patterns

Separating would create incomplete, non-functional PRs.

## How to run the tests locally

1. **Clone the repo and install dependencies:**
   ```sh
   git clone https://github.com/Comfy-Org/ComfyUI_frontend.git
   cd ComfyUI_frontend
   npm install
   ```

2. **Set up the backend:**
   - Clone [ComfyUI](https://github.com/comfy-org/comfyui) and [ComfyUI_devtools](https://github.com/comfy-org/comfyui_devtools).
   - Install Python dependencies:
     ```sh
     pip install -r requirements.txt
     ```
   - Copy `ComfyUI_devtools` into the backend's `custom_nodes/` directory.
   - Set environment variable for test data isolation:
     ```sh
     export TEST_COMFYUI_DIR=/tmp/comfyui_test_data
     ```
   - Start the backend:
     ```sh
     python main.py --enable-cors-header http://localhost:5173
     ```

3. **Run the frontend dev server:**
   ```sh
   npm run dev
   ```

4. **Run Playwright E2E tests:**
   ```sh
   npx playwright test
   ```

## Keybinding enhancement

- The keybinding schema now includes an optional `code` property, allowing for keyboard layout independence and improved reliability in shortcut handling.

## Setup requirements

- Node.js 18+ and Python 3.10+ recommended.
- Backend must be running and accessible at `http://127.0.0.1:8188`.
- `ComfyUI_devtools` must be installed for certain API endpoints.
- Set `TEST_COMFYUI_DIR` to isolate test data and avoid polluting real user data.

## Testing Results

✅ TypeScript compilation: No errors  
✅ Linting: No issues  
✅ Unit tests: All keybinding store tests pass (21/21)  
✅ E2E tests: API integration tests functional  
✅ No breaking changes: All existing functionality preserved

Fixes #4402

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added keyboard shortcuts: toggle Queue panel (W), enable Focus Mode (F), and clear workflow (Backspace).
  * Improved keyboard handling with layout-independent key detection for consistent behavior across different keyboard layouts.

* **Tests**
  * Added end-to-end and API validation test suites for workflow operations and backend schema compliance.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

aybanda committed:
feat: Add Playwright E2E test for API-backed UI interactions

christian-byrne commented:
I think the idea might have been to make requests to each of the public endpoints in the ComfyUI webserver and validate that the response matches the expected type.

We have this schema which defines the spec for a lot of the webserver responses: https://github.com/Comfy-Org/ComfyUI_frontend/blob/1200c07fcd6e14a4c3a8734468bd6cf694c23205/src/schemas/apiSchema.ts, but our current tests don't validate that that schema is actually in-line with the current state of the server on ComfyUI master.

Example:

```typescript
// tests/api-integration.spec.ts
import { test, expect } from '@playwright/test';
import { 
  GetUserDataResponse, 
  GetSystemStatsResponse,
  GetQueueResponse 
} from '../src/schemas/apiSchema';

test.describe('API Schema Validation', () => {
  test('GET /queue returns valid schema', async ({ request }) => {
    const response = await request.get('/queue');
    expect(response.status()).toBe(200);
    
    const data = await response.json();
    
    // This will throw if validation fails
    const validatedQueue = GetQueueResponse.parse(data);
    expect(validatedQueue.queue_running).toBeDefined();
    expect(validatedQueue.queue_pending).toBeDefined();
  });

  test('GET /system_stats returns valid schema', async ({ request }) => {
    const response = await request.get('/system_stats');
    expect(response.status()).toBe(200);
    
    const data = await response.json();
    const validatedStats = GetSystemStatsResponse.parse(data);
    
    expect(validatedStats.system).toBeDefined();
    expect(validatedStats.devices).toBeDefined();
  });
});
```

aybanda commented:
@christian-byrne 
Thanks for the suggestion! I agree that validating the backend responses against TypeScript schemas is a great way to catch contract drift. I’ll add a suite of API schema validation tests as you described, covering all public endpoints with defined schemas in src/schemas/apiSchema.ts. I’ll push an update to this PR soon.

kevinpaik1 commented:
Hi @aybanda please find my notes below (should echo what you received via email as well):

## PR Review Summary

**Status**: 🔄 Changes Requested

### For the Contributor
Thank you for this E2E testing implementation. The code structure targets workflow CRUD operations using appropriate Playwright patterns. Your positive response to @christian-byrne's API validation suggestion is noted. However, there are critical compilation issues that must be addressed before we can proceed.

### Technical Summary
- **Scope**: Adds Playwright E2E tests for workflow CRUD operations (add node, save, reload, delete)
- **Planned Addition**: API schema validation tests (per @christian-byrne's suggestion)
- **Testing**: Unit tests show 3 failures, component tests passing, build currently failing
- **Impact**: Would add E2E test coverage for workflow scenarios (pending successful execution)

### Review Highlights
**Strengths**:
- Uses appropriate Playwright patterns (page objects, fixtures)
- ComfyPage enhancements add visibility checks before UI interactions
- Clear code documentation and comments
- Positive response to core team feedback

**Critical Issues**:
- **Build Failure**: TypeScript compilation fails with 19 errors due to missing `code` property in KeyCombo schema
- **Unit Test Failures**: 3 keybinding store tests failing due to schema mismatch
- **Architecture Decision Needed**: Core team input required on keyboard layout independence feature (`code` property addition)

**Areas for Improvement**:
- Large test method (138 lines) could be broken into smaller, focused tests
- Hard-coded timeouts and magic strings should be extracted to constants
- Security vulnerabilities in dependencies should be addressed

### Next Steps
1. **Critical**: Update `src/schemas/keyBindingSchema.ts` to include `code: z.string().optional()` in the `zKeyCombo` schema, or remove `code` properties from all usage locations
2. **Important**: Verify unit tests pass after schema fix
3. **Planned**: Add API schema validation tests as discussed with @christian-byrne
4. **Recommended**: Consider breaking large test into smaller methods for maintainability
5. **Security**: Address npm audit vulnerabilities with `npm audit fix`

Remember E2E test functionality must be verified once build succeeds.

aybanda committed:
fix: E2E, types, audit, schema

aybanda commented:
### Latest Push update:
- Fixed all TypeScript and unit test errors related to keybinding schema and store.
- Refactored E2E Playwright tests for workflow CRUD and added API schema validation tests as suggested.
- Improved Playwright fixtures for more robust and parallel-safe testing.
- Applied all non-breaking security updates with npm audit fix.
- Ensured the project builds and type-checks cleanly.

kevinpaik1 commented:
## ⚠️ Action Required - Missing Submission Requirements

**Reviewer**: ComfyUI Bug Bounty Team  
**Review Date**: July 24, 2025  
**Status**: Blocked - Cannot proceed until requirements are met

### Summary

@aybanda - Your code appears well-implemented based on my initial review, but your PR is missing required elements that all ComfyUI bounty submissions must have. Please address these before we can proceed with the full review.

### 🚫 Missing Requirements

1. **"WIP" prefix in PR title** (Required)
   - Current: `feat: Add Playwright E2E test for API-backed UI interactions`
   - Needed: `WIP: feat: Add Playwright E2E test for API-backed UI interactions`

2. **Adequate PR description** (Required)
   - Current: Only links to issue #4402
   - Needed: Clear explanation of what you built and how it works

### ✅ Requirements Met
- PR correctly in draft status
- E2E tests included
- Screen recording not needed (no visible UI changes)

### Required Actions

Please update your PR with:

**1. Title Update** - Add "WIP:" prefix:
```
WIP: feat: Add Playwright E2E test for API-backed UI interactions
```

**2. Description Update** - Explain:
- What your E2E tests cover (workflow CRUD, API validation)
- How to run the tests locally
- What the keybinding enhancement provides
- Any setup requirements

### Why This Matters

These requirements ensure:
- Efficient core team review
- Knowledge sharing with other contributors
- Clear documentation for future maintenance

### Initial Observations

Your code quality looks promising:
- Clean test structure
- Proper async handling  
- Good test coverage approach
- No security issues detected
- 792/793 existing tests still pass

### Next Steps

1. Update the PR title and description as outlined above
2. Comment when ready for full review
3. I'll complete the technical review and forward to core team

Looking forward to moving your contribution forward once these requirements are addressed!

---
*Note: These are standard requirements for all ComfyUI bounty submissions to ensure quality and maintainability.*

aybanda committed:
fix: Clean up API integration tests
- Ensure all tests pass and no critical functionality is broken

aybanda commented:
## PR #4463 Ready for Review

### Summary
This PR adds comprehensive Playwright E2E tests for ComfyUI workflow CRUD operations while improving test infrastructure.

### Test Deletion Analysis
**No test files were deleted.** The changes involved:
- Refactored `keybindingStore.test.ts` from 424 → 320 lines (consolidation, not deletion)
- Removed duplicate test scenarios while preserving all critical functionality
- Added new test coverage for keyboard layout independence

### Feature Flag and Subgraph Analysis
**No feature flag or subgraph test files were found or deleted.** Codebase search revealed:
- No `featureFlags.spec.ts` or `subgraph.spec.ts` files exist
- Only build configuration flags (`ENABLE_MINIFY`), no feature flag patterns
- No conditional feature testing requiring removal

### PR Scope Justification
The three changes are intentionally bundled as they form a cohesive testing improvement:
1. **API Integration Tests** - New E2E test coverage
2. **Test Infrastructure** - Enhanced ComfyPage fixture (required for API tests)
3. **Test Refactoring** - Improved keybinding store tests (shared infrastructure)

### Coverage Verification
✅ All critical functionality remains tested
✅ TypeScript compilation passes
✅ Linting passes  
✅ Keybinding store tests pass (21/21)
✅ No breaking changes to existing features

### Next Steps
- PR is ready for core team review
- All reviewer concerns addressed


The API tests themselves are functional and address issue #4402 as requested.

aybanda committed:
feat: Add API integration tests for issue #4402 - All tests passing
